### PR TITLE
Enhanced Rdoc for Hash

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Covers methods:
- #flatten
- #compact
- #compact!
- #compare_by_identity
- #compare_by_identity?
- #any?
- #dig

### CODE REVIEW

**Comment by BurdetteLamar** [comment]
@drbrain, finally getting #dig moved over.

File: hash.c [hunk 1]
```diff
@@ -2506,7 +2506,7 @@ rb_hash_delete(VALUE hash, VALUE key)
 /*
  *  call-seq:
  *    hash.delete(key) -> value or nil
- *    hash.delete(key) { |key| ... } -> value
+ *    hash.delete(key) { |key| ... } -<value
```
Comment (nobu) [inline hash.c:2509]: A typo?

Comment (BurdetteLamar) [reply hash.c:2509]: That's affirmative.  Thanks.

File: hash.c [hunk 2]
```diff
@@ -4802,14 +4825,18 @@ rb_hash_compact(VALUE hash)
...
  *  call-seq:
- *     hsh.compact! -> hsh or nil
- *
- *  Removes all nil values from the hash.
- *  Returns nil if no changes were made, otherwise returns the hash.
+ *    hash.compact! -> self or nil
  *
- *     h = { a: 1, b: false, c: nil }
- *     h.compact!     #=> { a: 1, b: false }
+ *  Returns +self+ with all +nil+-valued entries removed:
```
Comment (drbrain) [inline hash.c:4830]: I think this should have a note that it modifies the object in-place to match the `h1.equal?(h)` example line

File: hash.c [hunk 3]
```diff
@@ -4830,18 +4857,32 @@ static st_table *rb_init_identtable_with_size(st_index_t size);
...
+ *  two keys are considered the same only if they are the same object;
+ *  returns +self+.
  *
- *     h1 = { "a" => 100, "b" => 200, :c => "c" }
- *     h1["a"]        #=> 100
- *     h1.compare_by_identity
- *     h1.compare_by_identity? #=> true
- *     h1["a".dup]    #=> nil  # different objects.
- *     h1[:c]         #=> "c"  # same symbols are all same.
+ *  By default, these two object are considered to be the same key,
```
Comment (drbrain) [inline hash.c:4866]: How about:

> By default, these two objects are considered to be the same key,
> So +s1+ will overwrite +s0+.

to fix "object" pluralization and be more explicit about the example below?

File: hash.c [hunk 4]
```diff
@@ -4830,18 +4857,32 @@ static st_table *rb_init_identtable_with_size(st_index_t size);
...
+ *    s1 = 'x'
+ *    s0.equal?(s0) # => false
+ *    h = {}
+ *    h.compare_by_identity? # => false
+ *    h[s0] = 0
+ *    h[s1] = 1
+ *    h # => {"x"=>1}
  *
+ *  After calling \#compare_by_identity, the keys are considered to be different,
+ *  and therefore do not overwrite:
```
Comment (drbrain) [inline hash.c:4878]: and finish this line with "… do not overwrite each other:"

**Comment by drbrain** [comment]
I think your PR description is a good commit description.

If you type a long commit description with markdown formatting and use it that single commit as a PR it will also become the PR description, and possibly save you some typing.

**Comment by BurdetteLamar** [comment]
> 
> 
> I think your PR description is a good commit description.
> 
> If you type a long commit description with markdown formatting and use it that single commit as a PR it will also become the PR description, and possibly save you some typing.

The good description is the title?  Or the bullet list?  I'm still not getting it about a mutliple-commit PR.

**Comment by drbrain** [comment]
> The good description is the title? Or the bullet list?

Both together. Here's a [recent commit](https://github.com/ruby/ruby/commit/b7e1de117e20911ed86b229a4c2f912e469c6472) that has a reasonable title and a few specifics in the body to give more context.  For this change, the current title and PR description together would be a good final commit message.

> I'm still not getting it about a mutliple-commit PR.

If you'd like we can schedule time for a zoom or other screen-sharing meeting to go over this more directly, perhaps be before merging this PR? Otherwise I can try to find some resources for you.

**Comment by BurdetteLamar** [comment]
If the description for this PR is ok, then I'm ok.  Thanks for help.
I think the recent complaint (well, not exactly a complaint) was b/c when I re-pushed, people got a message featuring "Respond to review," which is unhelpful?

File: hash.c [hunk 5]
```diff
@@ -4713,18 +4713,43 @@ flatten_i(VALUE key, VALUE val, VALUE ary)
...
+ *
+ *  Returns a new \Array object that is a 1-dimensional flattening of +self+.
+ *
+ *  ---
+ *
+ *  By default, nested Arrays are not flattened:
+ *    h = {foo: 0, bar: [:bat, 3], baz: 2}
+ *    h.flatten # => [:foo, 0, :bar, [:bat, 3], :baz, 2]
+ *
+ *  Takes the depth of recursive flattening from argument +level+:
```
Comment (ioquatix) [inline hash.c:4730]: It doesn't seem like a complete sentence.

**Pull request merged**